tsinfer/provenance.py

```diff
@@ -86,4 +90,51 @@ def get_provenance_dict(command=None, **kwargs):
         "parameters": parameters,
         "environment": get_environment(),
     }
+    if resources is not None:
+        document["resources"] = resources
     return document
+
+
+def get_peak_memory_bytes():
+    # peak memory usage in bytes
+    if sys.platform in ("linux", "darwin"):
+        usage = resource.getrusage(resource.RUSAGE_SELF)
+        max_rss = usage.ru_maxrss
+
+        if sys.platform == "darwin":
+            # macOS reports in bytes
+            return max_rss
+        else:
+            # Linux reports in kilobytes
+            return max_rss * 1024  # Convert KB to bytes
+
+    elif sys.platform == "win32":
+        return psutil.Process().memory_info().peak_wset
+
+    else:
+        return None
+
+
+class TimingAndMemory:
+    # Context manager for tracking timing and memory usage.
+    def __init__(self):
+        self.metrics = None
+
+    def __enter__(self):
+        self.start_process = psutil.Process()
+        self.start_elapsed = time.perf_counter()
+        self.start_times = self.start_process.cpu_times()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        end_times = self.start_process.cpu_times()
+        self.metrics = {
```

Comment:
Might as well make this a dataclass,
```
@dataclasses.dataclass
class ResourceMetrics:
     elapsed_time: float # etc
```
then, this can support aggregation, so you can sum over a bunch of them.

Can add the ``asdict()`` method for getting out the dict, as usual.